llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -5286,6 +5286,106 @@ MachineInstr *CombinerHelper::buildSDivUsingMul(MachineInstr &MI) {
   return MIB.buildMul(Ty, Res, Factor);
 }
 
+bool CombinerHelper::matchSDivByPow2(MachineInstr &MI) {
+  assert(MI.getOpcode() == TargetOpcode::G_SDIV && "Expected SDIV");
+  if (MI.getFlag(MachineInstr::MIFlag::IsExact))
+    return false;
+  auto &SDiv = cast<GenericMachineInstr>(MI);
+  Register RHS = SDiv.getReg(2);
+  auto MatchPow2 = [&](const Constant *C) {
+    if (auto *CI = dyn_cast<ConstantInt>(C))
+      return CI->getValue().isPowerOf2() || CI->getValue().isNegatedPowerOf2();
+    return false;
+  };
+  return matchUnaryPredicate(MRI, RHS, MatchPow2, /* AllowUndefs */ false);
+}
+
+void CombinerHelper::applySDivByPow2(MachineInstr &MI) {
+  assert(MI.getOpcode() == TargetOpcode::G_SDIV && "Expected SDIV");
+  auto &SDiv = cast<GenericMachineInstr>(MI);
+  Register Dst = SDiv.getReg(0);
+  Register LHS = SDiv.getReg(1);
+  Register RHS = SDiv.getReg(2);
+  LLT Ty = MRI.getType(Dst);
+  LLT ShiftAmtTy = getTargetLowering().getPreferredShiftAmountTy(Ty);
+
+  Builder.setInstrAndDebugLoc(MI);
+
+  auto RHSC = getIConstantVRegValWithLookThrough(RHS, MRI);
```

Comment:
It looks like this will not work when `RHS` is a vector.
```
define <2 x i128> @v_sdiv_v2i128(<2 x i128> %num) {
  %result = sdiv <2 x i128> %num, <i128 8589934592, i128 4096>
  ret <2 x i128> %result
}
```
We might still need to build the instructions and then fold them accordingly.

Reply:
We're lacking good vector handling, and we don't really handle non-splat vectors anywhere In globalisel. We need some new helpers for this

Reply:
For simple cases, you can try:
```
std::optional<APInt> MaybeRHS = getConstantOrConstantSplatVector(RHS);
```

Reply:
```
std::optional<SmallVector<APInt>> = getConstantOrConstantVector(RHS, **UpperBound**);
```
X86 has vectors up to 512 bits. There are publicly known Arm cores with 512 bit vectors. This will seriously need an upper bound.